File: service/src/evidencegraph_service/protocol.py
"""Private-channel wire protocol (S1-04).

Tauri (the parent) and this service's supervised-mode child
(`supervised.py`) exchange a small, fixed-order sequence of framed JSON
messages over the child's inherited stdin (parent -> child) and stdout
(child -> parent). Structured logs continue to go to stderr
(`logging_config.py`), a completely separate stream -- protocol messages
and log lines can never be confused with each other because they never
share a stream, and this module never writes anything to `stdout` except
the exact protocol frames defined here (nothing here calls `print`, and
`write_frame` is the only function that touches the child's stdout at
all).

Wire format (must match `app/src-tauri/src/supervisor/protocol.rs`
byte-for-byte):

    frame = u32_big_endian(len(payload)) ++ payload
    payload = UTF-8-encoded JSON object, always containing at least
              {"v": 1, "type": "<message-type>"}

`len(payload)` is bounded by `MAX_FRAME_BYTES`; a frame claiming to be
larger is a protocol violation, not a value to attempt reading. Every
message declares protocol version 1 as a JSON *integer* (`"v": 1`) -- a
JSON boolean or float in that position is rejected the same as a wrong
integer (Python's `bool` is a subtype of `int`, so this module always
checks `type(value) is int`, never `isinstance(value, int)`, to keep
`true`/`false` from silently passing as `1`/`0`). A JSON object is parsed
with duplicate keys rejected outright (see `_strict_object_pairs_hook`) -- never
silently collapsed to last-value-wins -- and, per message type, only the
exact expected key set is accepted; an extra or missing key is a protocol
violation.

Message order (see `docs/SPRINT_1_BACKLOG.md` S1-04 and
`docs/DECISIONS.md` D-025 for the full handshake this participates in):

    parent -> child   startup_secret          (must be the first message)
    child  -> parent  endpoint_ready
    ... (D-025 HTTP challenge/response happens out-of-band, over the
         bound HTTP endpoint, not this channel -- see `challenge.py`) ...
    parent -> child   install_session_token
    child  -> parent  ready
    parent -> child   shutdown                (optional, at app exit)
    child  -> parent  shutdown_ack

Any message out of this order, any duplicate, any unknown `type`, EOF
before `ready`, or a malformed/oversized frame is fatal to the child
process (`ProtocolError`, uncaught by design at the call site in
`supervised.py`, which exits non-zero) -- "unknown, malformed, duplicated,
and out-of-order messages fail startup."

Every `ProtocolError` this module raises carries one of a small, fixed set
of category strings, defined once as `_Reason` constants below -- never an
f-string built from the received message's `type`, field contents, or any
other attacker/parent-controlled value. This is deliberate: these
messages can end up in structured logs (`supervised.py`) and, indirectly,
in the frontend-visible `Failed` reason on the Rust side -- neither may
ever echo back received content.
"""

from __future__ import annotations

import json
import struct
from typing import Protocol

from . import b64url

PROTOCOL_VERSION = 1

#: Real messages here are tiny (a handful of short fields); this bound
#: exists purely so a malformed/hostile length prefix can never make the
#: child attempt to allocate or read an unbounded amount of data.
MAX_FRAME_BYTES = 4096

#: Exact required decoded length of the startup secret -- part of S1-04's
#: strict wire schema. Matches the 32 random bytes Tauri generates
#: (`app/src-tauri/src/supervisor/mod.rs::STARTUP_SECRET_LEN`).
STARTUP_SECRET_LENGTH = 32

_LENGTH_STRUCT = struct.Struct(">I")  # big-endian uint32

# --- Fixed, bounded error reasons -------------------------------------------
# Every one of these is a literal string constant. Nothing in this module
# ever builds a `ProtocolError` from `f"...{attacker_or_parent_value}..."`.

_REASON_CHANNEL_CLOSED = "channel_closed"
_REASON_FRAME_TOO_LARGE = "frame_too_large"
_REASON_FRAME_LENGTH_OUT_OF_BOUNDS = "frame_length_out_of_bounds"
_REASON_FRAME_NOT_VALID_JSON = "frame_not_valid_json"
_REASON_FRAME_DUPLICATE_KEY = "frame_duplicate_key"
_REASON_FRAME_NOT_AN_OBJECT = "frame_not_an_object"
_REASON_UNSUPPORTED_PROTOCOL_VERSION = "unsupported_protocol_version"
_REASON_FRAME_MISSING_TYPE = "frame_missing_type"
_REASON_UNEXPECTED_MESSAGE_TYPE = "unexpected_message_type"
_REASON_UNEXPECTED_SCHEMA = "unexpected_schema"


class ProtocolError(Exception):
    """Any private-channel framing/ordering/content violation. Always
    fatal -- there is no partial-recovery path. The message is always one
    of the fixed `_REASON_*` category strings above -- never request
    content, a received message type, a field value, or a path."""


class BinaryStream(Protocol):
    """Structural type for the raw stdin/stdout buffer objects this module
    operates on (`sys.stdin.buffer` / `sys.stdout.buffer` in production;
    `io.BytesIO` in tests)."""

    def read(self, n: int, /) -> bytes: ...
    def write(self, data: bytes, /) -> int: ...
    def flush(self) -> None: ...


def _strict_object_pairs_hook(pairs: list[tuple[str, object]]) -> dict[str, object]:
    """`json.loads(..., object_pairs_hook=...)` sees every key exactly as
    it appeared in the input, before the standard dict-building step would
    silently collapse duplicates (last-value-wins). Raising here is what
    makes a duplicate key a parse failure instead of a silently-accepted
    ambiguity."""
    seen: dict[str, object] = {}
    for key, value in pairs:
        if key in seen:
            raise ValueError(_REASON_FRAME_DUPLICATE_KEY)
        seen[key] = value
    return seen


def write_frame(stream: BinaryStream, message: dict[str, object]) -> None:
    payload = json.dumps(message, separators=(",", ":")).encode("utf-8")
    if len(payload) > MAX_FRAME_BYTES:
        raise ProtocolError(_REASON_FRAME_TOO_LARGE)
    stream.write(_LENGTH_STRUCT.pack(len(payload)))
    stream.write(payload)
    stream.flush()


def _read_exact(stream: BinaryStream, n: int) -> bytes:
    chunks = bytearray()
    while len(chunks) < n:
        chunk = stream.read(n - len(chunks))
        if not chunk:
            raise ProtocolError(_REASON_CHANNEL_CLOSED)
        chunks.extend(chunk)
    return bytes(chunks)


def read_frame(stream: BinaryStream) -> dict[str, object]:
    """Reads one frame and strictly validates its envelope: a JSON object
    with no duplicate keys, an integer (never bool/float) `"v" ==
    PROTOCOL_VERSION`, and a string `"type"`. Does **not** check the
    exact key set -- that depends on which message type this turns out to
    be, which the caller doesn't know until after this returns; callers
    must follow up with `_require_exact_keys` once they've routed on
    `message["type"]`.
    """
    (length,) = _LENGTH_STRUCT.unpack(_read_exact(stream, _LENGTH_STRUCT.size))
    if length == 0 or length > MAX_FRAME_BYTES:
        raise ProtocolError(_REASON_FRAME_LENGTH_OUT_OF_BOUNDS)
    payload = _read_exact(stream, length)
    try:
        message = json.loads(payload.decode("utf-8"), object_pairs_hook=_strict_object_pairs_hook)
    except (UnicodeDecodeError, ValueError) as exc:
        raise ProtocolError(_REASON_FRAME_NOT_VALID_JSON) from exc
    if not isinstance(message, dict):
        raise ProtocolError(_REASON_FRAME_NOT_AN_OBJECT)
    version = message.get("v")
    if type(version) is not int or version != PROTOCOL_VERSION:
        # `type(...) is not int` (not `isinstance`) deliberately excludes
        # `bool` -- `True == 1` in Python, so `isinstance(True, int) and
        # True == 1` would incorrectly accept `{"v": true}`.
        raise ProtocolError(_REASON_UNSUPPORTED_PROTOCOL_VERSION)
    if not isinstance(message.get("type"), str):
        raise ProtocolError(_REASON_FRAME_MISSING_TYPE)
    return message


def _require_exact_keys(message: dict[str, object], expected: frozenset[str]) -> None:
    if set(message.keys()) != expected:
        raise ProtocolError(_REASON_UNEXPECTED_SCHEMA)


# --- Typed helpers for the child's side of the exchange -------------------

_STARTUP_SECRET_KEYS = frozenset({"v", "type", "secret"})
_INSTALL_TOKEN_KEYS = frozenset({"v", "type", "token"})
_SHUTDOWN_KEYS = frozenset({"v", "type"})


def read_startup_secret(stream: BinaryStream) -> bytes:
    """Reads and strictly validates the mandatory first message. Anything
    else -- wrong type, missing/malformed/wrong-length `secret` field,
    unknown/missing keys, EOF -- is a `ProtocolError`, and the caller must
    never open a socket first."""
    message = read_frame(stream)
    if message["type"] != "startup_secret":
        raise ProtocolError(_REASON_UNEXPECTED_MESSAGE_TYPE)
    _require_exact_keys(message, _STARTUP_SECRET_KEYS)
    secret_b64 = message.get("secret")
    try:
        return b64url.decode_exact(
            secret_b64, field="startup_secret", expected_length=STARTUP_SECRET_LENGTH
        )
    except b64url.DecodeError as exc:
        raise ProtocolError(str(exc)) from exc


def write_endpoint_ready(stream: BinaryStream, *, host: str, port: int) -> None:
    write_frame(
        stream, {"v": PROTOCOL_VERSION, "type": "endpoint_ready", "host": host, "port": port}
    )


def read_install_token_or_shutdown(stream: BinaryStream) -> tuple[str, str | None]:
    """Reads exactly one message from the two types valid at this point in
    the sequence. Returns `("install_session_token", token)` or
    `("shutdown", None)`. Any other `type` -- including a second
    `startup_secret` -- is a `ProtocolError`: this channel's message order
    is fixed, not a general-purpose queue. The message is routed on
    `type` first, then checked against *that* type's own exact key set --
    an `install_session_token` with an extra field, or a `shutdown` with
    any field beyond `v`/`type`, is rejected either way.
    """
    message = read_frame(stream)
    msg_type = message["type"]
    if not isinstance(msg_type, str):
        raise ProtocolError(_REASON_FRAME_MISSING_TYPE)  # unreachable: read_frame already checked
    if msg_type == "install_session_token":
        _require_exact_keys(message, _INSTALL_TOKEN_KEYS)
        token = message.get("token")
        if not isinstance(token, str) or not token:
            raise ProtocolError(_REASON_UNEXPECTED_SCHEMA)
        return msg_type, token
    if msg_type == "shutdown":
        _require_exact_keys(message, _SHUTDOWN_KEYS)
        return msg_type, None
    raise ProtocolError(_REASON_UNEXPECTED_MESSAGE_TYPE)


def write_ready(stream: BinaryStream) -> None:
    write_frame(stream, {"v": PROTOCOL_VERSION, "type": "ready"})


def write_shutdown_ack(stream: BinaryStream) -> None:
    write_frame(stream, {"v": PROTOCOL_VERSION, "type": "shutdown_ack"})


#: Small, fixed set of reasons the child may report before it can even
#: install a credential (e.g. a malformed `startup_secret` message) --
#: never a raw exception message, which could echo back attacker/parent-
#: controlled content.
_ERROR_REASONS = frozenset(
    {
        "invalid_startup_secret",
        "bind_failed",
        "internal_error",
    }
)


def write_error(stream: BinaryStream, *, reason: str) -> None:
    if reason not in _ERROR_REASONS:
        reason = "internal_error"
    write_frame(stream, {"v": PROTOCOL_VERSION, "type": "error", "reason": reason})

File: app/src-tauri/src/supervisor/protocol.rs
//! Private-channel wire protocol (S1-04). Mirrors
//! `service/src/evidencegraph_service/protocol.py` byte-for-byte -- see
//! that module's docstring for the full framing/ordering contract this
//! implements. Only framing and the specific message shapes live here;
//! message *ordering* is enforced by the caller (`supervisor/mod.rs`),
//! which reads and writes messages in the one fixed sequence the
//! handshake requires -- there is no general-purpose message queue here.
//!
//! Every `ProtocolError` variant here is a fixed, content-free tag --
//! never a received message's `type`, a field value, or any other
//! child-controlled content. An earlier version of this module carried
//! `UnexpectedType(String)`/`MissingField(&'static str)` payloads and
//! interpolated the received type into `Display`, which is exactly the
//! class of leak S1-04 review finding 4 flagged; every call site in
//! `supervisor::mod` now maps a `ProtocolError` to a
//! [`super::errors::FailureReason`] before it can reach a log or
//! `SupervisorState::Failed`, so nothing here needs -- or is permitted --
//! to carry request content to do that mapping usefully.
//!
//! Objects are parsed with [`parse_object_rejecting_duplicates`], never
//! plain `serde_json::from_slice::<Value>`: the latter silently collapses
//! a duplicate key to its last occurrence, which is exactly the kind of
//! ambiguity a strict wire schema (S1-04 review finding 5) must not
//! accept silently.

use super::b64url;
use serde::de::{Deserializer as _, Error as DeError, MapAccess, Visitor};
use serde_json::{Map, Value};
use std::fmt;
use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};

pub const PROTOCOL_VERSION: u64 = 1;

/// Matches `protocol.MAX_FRAME_BYTES` in the Python implementation.
pub const MAX_FRAME_BYTES: u32 = 4096;

/// Matches `protocol.STARTUP_SECRET_LENGTH`/`config._REQUIRED_HOST` on the
/// Python side and `supervisor::STARTUP_SECRET_LEN`/`process::spawn`'s own
/// bind contract on this side.
const REQUIRED_HOST: &str = "127.0.0.1";

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProtocolError {
    ChannelClosed,
    FrameTooLarge,
    FrameLengthOutOfBounds,
    FrameNotValidJson,
    /// Valid JSON, but not an object, or an object with a duplicate key --
    /// deliberately not distinguished further (both are "the framing
    /// contract was violated before we ever got to field-level checks").
    FrameMalformedObject,
    UnsupportedProtocolVersion,
    FrameMissingType,
    UnexpectedMessageType,
    /// Missing/extra/wrong-typed field, or a field value outside its
    /// required range/shape (e.g. a non-loopback `host`, an out-of-range
    /// `port`).
    UnexpectedSchema,
    HostNotLoopback,
    Io,
}

impl fmt::Display for ProtocolError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        // Every branch is a fixed literal -- see module docs.
        let s = match self {
            Self::ChannelClosed => "channel_closed",
            Self::FrameTooLarge => "frame_too_large",
            Self::FrameLengthOutOfBounds => "frame_length_out_of_bounds",
            Self::FrameNotValidJson => "frame_not_valid_json",
            Self::FrameMalformedObject => "frame_malformed_object",
            Self::UnsupportedProtocolVersion => "unsupported_protocol_version",
            Self::FrameMissingType => "frame_missing_type",
            Self::UnexpectedMessageType => "unexpected_message_type",
            Self::UnexpectedSchema => "unexpected_schema",
            Self::HostNotLoopback => "host_not_loopback",
            Self::Io => "private_channel_io_error",
        };
        f.write_str(s)
    }
}

impl std::error::Error for ProtocolError {}

impl From<std::io::Error> for ProtocolError {
    fn from(_: std::io::Error) -> Self {
        // The underlying `io::Error` is deliberately discarded here, not
        // wrapped -- its `Display` can include OS-provided text that
        // isn't one of this module's fixed categories.
        Self::Io
    }
}

/// A `serde` `Visitor` that parses a JSON object while rejecting any
/// duplicate key -- `serde_json::Value`'s normal map-building silently
/// keeps only the *last* occurrence of a repeated key, which is exactly
/// the ambiguity a strict wire schema must not accept. Any non-object
/// top-level value also fails here (via the `Visitor` trait's default
/// `visit_*` methods, which we don't override), collapsed into the same
/// [`ProtocolError::FrameMalformedObject`] as a duplicate key -- both are
/// "the object-level framing contract was violated."
struct NoDuplicateKeysVisitor;

impl<'de> Visitor<'de> for NoDuplicateKeysVisitor {
    type Value = Map<String, Value>;

    fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("a JSON object with no duplicate keys")
    }

    fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
    where
        A: MapAccess<'de>,
    {
        let mut result = Map::new();
        while let Some((key, value)) = map.next_entry::<String, Value>()? {
            if result.contains_key(&key) {
                return Err(A::Error::custom("duplicate key"));
            }
            result.insert(key, value);
        }
        Ok(result)
    }
}

fn parse_object_rejecting_duplicates(bytes: &[u8]) -> Result<Map<String, Value>, ProtocolError> {
    let mut deserializer = serde_json::Deserializer::from_slice(bytes);
    match deserializer.deserialize_any(NoDuplicateKeysVisitor) {
        Ok(map) => Ok(map),
        Err(e) if e.is_syntax() || e.is_eof() => Err(ProtocolError::FrameNotValidJson),
        Err(_) => Err(ProtocolError::FrameMalformedObject),
    }
}

pub async fn write_frame<W: AsyncWrite + Unpin>(
    writer: &mut W,
    message: &Value,
) -> Result<(), ProtocolError> {
    let payload = serde_json::to_vec(message).map_err(|_| ProtocolError::FrameNotValidJson)?;
    if payload.len() as u32 > MAX_FRAME_BYTES {
        return Err(ProtocolError::FrameTooLarge);
    }
    writer.write_u32(payload.len() as u32).await?;
    writer.write_all(&payload).await?;
    writer.flush().await?;
    Ok(())
}

/// Reads one frame and strictly validates its envelope: a JSON object
/// with no duplicate keys, an integer (never bool/float -- `Value::
/// as_u64` returns `None` for both JSON variants, so this is already
/// correct without an extra check) `"v" == PROTOCOL_VERSION`, and a
/// string `"type"`. Does **not** check the exact key set -- that depends
/// on which message type this turns out to be, which the caller doesn't
/// know until after this returns; callers must follow up with
/// [`require_exact_keys`] once they've routed on the `"type"` value.
async fn read_frame<R: AsyncRead + Unpin>(
    reader: &mut R,
) -> Result<Map<String, Value>, ProtocolError> {
    let length = reader
        .read_u32()
        .await
        .map_err(|_| ProtocolError::ChannelClosed)?;
    if length == 0 || length > MAX_FRAME_BYTES {
        return Err(ProtocolError::FrameLengthOutOfBounds);
    }
    let mut buf = vec![0u8; length as usize];
    reader
        .read_exact(&mut buf)
        .await
        .map_err(|_| ProtocolError::ChannelClosed)?;

    let message = parse_object_rejecting_duplicates(&buf)?;

    if message.get("v").and_then(Value::as_u64) != Some(PROTOCOL_VERSION) {
        Err(ProtocolError::UnsupportedProtocolVersion)
    } else if !message.get("type").is_some_and(Value::is_string) {
        Err(ProtocolError::FrameMissingType)
    } else {
        Ok(message)
    }
}

fn require_exact_keys(
    message: &Map<String, Value>,
    expected: &[&str],
) -> Result<(), ProtocolError> {
    if message.len() != expected.len() || !expected.iter().all(|k| message.contains_key(*k)) {
        return Err(ProtocolError::UnexpectedSchema);
    }
    Ok(())
}

fn message_type(value: &Map<String, Value>) -> &str {
    value["type"].as_str().unwrap_or("")
}

// --- Typed helpers for the parent's side of the exchange -------------------

pub async fn write_startup_secret<W: AsyncWrite + Unpin>(
    writer: &mut W,
    secret: &[u8],
) -> Result<(), ProtocolError> {
    write_frame(
        writer,
        &serde_json::json!({"v": PROTOCOL_VERSION, "type": "startup_secret", "secret": b64url::encode(secret)}),
    )
    .await
}

pub struct EndpointReady {
    pub host: String,
    pub port: u16,
}

/// Reads and strictly validates the mandatory first child->parent
/// message. Anything else -- wrong type, missing/extra/wrong-typed
/// fields, a non-loopback `host`, an out-of-range `port`, EOF -- is a
/// `ProtocolError`, and the caller must never treat the child as
/// trustworthy first. `host` is checked against [`REQUIRED_HOST`]
/// *here*, at parse time (S1-04 review finding 5's "require host exactly
/// 127.0.0.1"), not left to a later business-logic check the caller could
/// forget -- a successfully returned `EndpointReady` is guaranteed
/// loopback by construction, so callers never need to re-check it.
pub async fn read_endpoint_ready<R: AsyncRead + Unpin>(
    reader: &mut R,
) -> Result<EndpointReady, ProtocolError> {
    let message = read_frame(reader).await?;
    if message_type(&message) != "endpoint_ready" {
        return Err(ProtocolError::UnexpectedMessageType);
    }
    require_exact_keys(&message, &["v", "type", "host", "port"])?;

    let host = message
        .get("host")
        .and_then(Value::as_str)
        .ok_or(ProtocolError::UnexpectedSchema)?;
    if host != REQUIRED_HOST {
        return Err(ProtocolError::HostNotLoopback);
    }

    let port = message
        .get("port")
        .and_then(Value::as_u64)
        .ok_or(ProtocolError::UnexpectedSchema)?;
    let port = u16::try_from(port).map_err(|_| ProtocolError::UnexpectedSchema)?;
    if port == 0 {
        return Err(ProtocolError::UnexpectedSchema);
    }

    Ok(EndpointReady {
        host: host.to_string(),
        port,
    })
}

pub async fn write_install_session_token<W: AsyncWrite + Unpin>(
    writer: &mut W,
    token: &str,
) -> Result<(), ProtocolError> {
    write_frame(
        writer,
        &serde_json::json!({"v": PROTOCOL_VERSION, "type": "install_session_token", "token": token}),
    )
    .await
}

pub async fn read_ready<R: AsyncRead + Unpin>(reader: &mut R) -> Result<(), ProtocolError> {
    let message = read_frame(reader).await?;
    if message_type(&message) != "ready" {
        return Err(ProtocolError::UnexpectedMessageType);
    }
    require_exact_keys(&message, &["v", "type"])
}

pub async fn write_shutdown<W: AsyncWrite + Unpin>(writer: &mut W) -> Result<(), ProtocolError> {
    write_frame(
        writer,
        &serde_json::json!({"v": PROTOCOL_VERSION, "type": "shutdown"}),
    )
    .await
}

pub async fn read_shutdown_ack<R: AsyncRead + Unpin>(reader: &mut R) -> Result<(), ProtocolError> {
    let message = read_frame(reader).await?;
    if message_type(&message) != "shutdown_ack" {
        return Err(ProtocolError::UnexpectedMessageType);
    }
    require_exact_keys(&message, &["v", "type"])
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Cursor;

    async fn write_raw(buf: &mut Vec<u8>, message: &Value) {
        write_frame(buf, message).await.unwrap();
    }

    #[tokio::test]
    async fn write_then_read_endpoint_ready_round_trips() {
        let mut buf = Vec::new();
        write_raw(
            &mut buf,
            &serde_json::json!({"v": 1, "type": "endpoint_ready", "host": "127.0.0.1", "port": 54321}),
        )
        .await;
        let mut cursor = Cursor::new(buf);
        let endpoint = read_endpoint_ready(&mut cursor).await.unwrap();
        assert_eq!(endpoint.host, "127.0.0.1");
        assert_eq!(endpoint.port, 54321);
    }

    #[tokio::test]
    async fn read_endpoint_ready_rejects_wrong_type() {
        let mut buf = Vec::new();
        write_raw(&mut buf, &serde_json::json!({"v": 1, "type": "ready"})).await;
        let mut cursor = Cursor::new(buf);
        assert!(read_endpoint_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn read_frame_rejects_eof() {
        let mut cursor = Cursor::new(Vec::<u8>::new());
        assert!(read_endpoint_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn read_frame_rejects_oversized_length_prefix() {
        let mut buf = Vec::new();
        buf.extend_from_slice(&(MAX_FRAME_BYTES + 1).to_be_bytes());
        let mut cursor = Cursor::new(buf);
        assert!(read_endpoint_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn read_frame_rejects_wrong_version() {
        let mut buf = Vec::new();
        write_raw(
            &mut buf,
            &serde_json::json!({"v": 2, "type": "endpoint_ready", "host": "127.0.0.1", "port": 1}),
        )
        .await;
        let mut cursor = Cursor::new(buf);
        assert!(read_endpoint_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn read_frame_rejects_non_object() {
        let mut buf = Vec::new();
        let payload = serde_json::to_vec(&serde_json::json!([1, 2, 3])).unwrap();
        buf.extend_from_slice(&(payload.len() as u32).to_be_bytes());
        buf.extend_from_slice(&payload);
        let mut cursor = Cursor::new(buf);
        assert!(read_endpoint_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn write_frame_rejects_oversized_message() {
        let huge = serde_json::json!({"v": 1, "type": "ready", "padding": "x".repeat(MAX_FRAME_BYTES as usize * 2)});
        let mut buf = Vec::new();
        assert!(write_frame(&mut buf, &huge).await.is_err());
    }

    #[tokio::test]
    async fn read_ready_and_shutdown_ack_shapes() {
        let mut buf = Vec::new();
        write_raw(&mut buf, &serde_json::json!({"v": 1, "type": "ready"})).await;
        write_raw(
            &mut buf,
            &serde_json::json!({"v": 1, "type": "shutdown_ack"}),
        )
        .await;
        let mut cursor = Cursor::new(buf);
        read_ready(&mut cursor).await.unwrap();
        read_shutdown_ack(&mut cursor).await.unwrap();
    }

    #[tokio::test]
    async fn write_startup_secret_and_install_session_token_are_valid_base64url() {
        let mut buf = Vec::new();
        write_startup_secret(&mut buf, &[1, 2, 3, 4]).await.unwrap();
        write_install_session_token(&mut buf, "some-token-value-here-1234567890")
            .await
            .unwrap();
        let mut cursor = Cursor::new(buf);
        let secret_msg = read_frame(&mut cursor).await.unwrap();
        assert_eq!(secret_msg["type"], "startup_secret");
        let secret_b64 = secret_msg["secret"].as_str().unwrap();
        assert_eq!(b64url::decode(secret_b64).unwrap(), vec![1, 2, 3, 4]);

        let token_msg = read_frame(&mut cursor).await.unwrap();
        assert_eq!(token_msg["type"], "install_session_token");
        assert_eq!(token_msg["token"], "some-token-value-here-1234567890");
    }

    // --- S1-04 review finding 5: strict wire schema -------------------------

    #[tokio::test]
    async fn read_frame_rejects_duplicate_v_key() {
        let mut buf = Vec::new();
        let payload = b"{\"v\":1,\"v\":1,\"type\":\"ready\"}";
        buf.extend_from_slice(&(payload.len() as u32).to_be_bytes());
        buf.extend_from_slice(payload);
        let mut cursor = Cursor::new(buf);
        assert!(read_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn read_frame_rejects_duplicate_key_in_a_typed_field() {
        let mut buf = Vec::new();
        let payload =
            b"{\"v\":1,\"type\":\"endpoint_ready\",\"host\":\"127.0.0.1\",\"host\":\"127.0.0.1\",\"port\":1}";
        buf.extend_from_slice(&(payload.len() as u32).to_be_bytes());
        buf.extend_from_slice(payload);
        let mut cursor = Cursor::new(buf);
        assert!(read_endpoint_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn read_frame_rejects_boolean_version() {
        for version in [serde_json::json!(true), serde_json::json!(false)] {
            let mut buf = Vec::new();
            write_raw(
                &mut buf,
                &serde_json::json!({"v": version, "type": "ready"}),
            )
            .await;
            let mut cursor = Cursor::new(buf);
            assert!(read_ready(&mut cursor).await.is_err());
        }
    }

    #[tokio::test]
    async fn read_frame_rejects_float_version() {
        let mut buf = Vec::new();
        write_raw(&mut buf, &serde_json::json!({"v": 1.0, "type": "ready"})).await;
        let mut cursor = Cursor::new(buf);
        assert!(read_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn read_frame_rejects_string_version() {
        let mut buf = Vec::new();
        write_raw(&mut buf, &serde_json::json!({"v": "1", "type": "ready"})).await;
        let mut cursor = Cursor::new(buf);
        assert!(read_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn read_endpoint_ready_rejects_non_loopback_host() {
        for host in ["localhost", "0.0.0.0", "::1", "192.168.1.5", "127.0.0.2"] {
            let mut buf = Vec::new();
            write_raw(
                &mut buf,
                &serde_json::json!({"v": 1, "type": "endpoint_ready", "host": host, "port": 1}),
            )
            .await;
            let mut cursor = Cursor::new(buf);
            assert!(
                read_endpoint_ready(&mut cursor).await.is_err(),
                "host {host} should be rejected"
            );
        }
    }

    #[tokio::test]
    async fn read_endpoint_ready_rejects_port_zero() {
        let mut buf = Vec::new();
        write_raw(
            &mut buf,
            &serde_json::json!({"v": 1, "type": "endpoint_ready", "host": "127.0.0.1", "port": 0}),
        )
        .await;
        let mut cursor = Cursor::new(buf);
        assert!(read_endpoint_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn read_endpoint_ready_rejects_port_above_65535() {
        let mut buf = Vec::new();
        write_raw(
            &mut buf,
            &serde_json::json!({"v": 1, "type": "endpoint_ready", "host": "127.0.0.1", "port": 65536}),
        )
        .await;
        let mut cursor = Cursor::new(buf);
        assert!(read_endpoint_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn read_endpoint_ready_rejects_boolean_port() {
        let mut buf = Vec::new();
        write_raw(
            &mut buf,
            &serde_json::json!({"v": 1, "type": "endpoint_ready", "host": "127.0.0.1", "port": true}),
        )
        .await;
        let mut cursor = Cursor::new(buf);
        assert!(read_endpoint_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn read_endpoint_ready_rejects_unknown_extra_key() {
        let mut buf = Vec::new();
        write_raw(
            &mut buf,
            &serde_json::json!({"v": 1, "type": "endpoint_ready", "host": "127.0.0.1", "port": 1, "extra": "x"}),
        )
        .await;
        let mut cursor = Cursor::new(buf);
        assert!(read_endpoint_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn read_ready_rejects_unknown_extra_key() {
        let mut buf = Vec::new();
        write_raw(
            &mut buf,
            &serde_json::json!({"v": 1, "type": "ready", "extra": "x"}),
        )
        .await;
        let mut cursor = Cursor::new(buf);
        assert!(read_ready(&mut cursor).await.is_err());
    }

    #[tokio::test]
    async fn multiple_frames_can_be_read_sequentially() {
        let mut buf = Vec::new();
        write_startup_secret(&mut buf, &[1, 2, 3, 4]).await.unwrap();
        write_raw(
            &mut buf,
            &serde_json::json!({"v": 1, "type": "endpoint_ready", "host": "127.0.0.1", "port": 1234}),
        )
        .await;
        write_install_session_token(&mut buf, "some-token-value-here-1234567890")
            .await
            .unwrap();
        write_raw(&mut buf, &serde_json::json!({"v": 1, "type": "ready"})).await;
        let mut cursor = Cursor::new(buf);

        let secret_msg = read_frame(&mut cursor).await.unwrap();
        assert_eq!(secret_msg["type"], "startup_secret");
        let endpoint = read_endpoint_ready(&mut cursor).await.unwrap();
        assert_eq!(endpoint.port, 1234);
        let token_msg = read_frame(&mut cursor).await.unwrap();
        assert_eq!(token_msg["type"], "install_session_token");
        read_ready(&mut cursor).await.unwrap();
    }
}
